src/background/hypothesis-chrome-extension.js

```diff
@@ -120,14 +120,28 @@ export default function HypothesisChromeExtension({
     });
   }
 
+  /**
+   *
+   * @param {number} tabId
+   * @param {import('./tab-state').State|undefined} current
+   */
   function onTabStateChange(tabId, current) {
     if (current) {
-      browserAction.update(tabId, current);
-      chromeTabs.get(tabId, updateTabDocument);
+      chromeTabs.get(tabId, tab => {
```

Comment:
I wondered here whether higher-level / earler code had made an error if we call `chromeTabs.get` for a tab that doesn't exist. From what I can see the sequence of events that happens when logging in via the Hypothesis client's authorization popup is:

1. User submits the form
2. `onTabUpdated` is called for the popup window with `{status: "complete"}` in the `changeInfo`. This updates the tab's state.
3. The state update triggers a call to `onTabStateChange` which in turn calls `chrome.tabs.get(...)`
4. `onTabRemoved` is called for the popup window, which removes the tab's state from the local store
5. The async callback for the `chrome.tabs.get(...)` call in step 3 fires.

So in this case the tab ID was valid at the point where `chrome.tabs.get` was called, but not when the callback got processed.

I think it would be worth noting in the comment you added below that `chrome.tabs.get` can fail even if the tab ID was valid (or at least, we hadn't received a notification yet that the tab had been removed) when `chrome.tabs.get` was called.

Reply:
I don't think the ID is valid when `chrome.tabs.get is called`. 

There are a few more callbacks fired:
6. The async callback after the badge count is fetched (which is now delayed by at least 500ms).
7. The async callback after calling `onTabRemoved` on step 4 (which triggers a state update).

This recursive callbacks are not ideal. I would suggest to change this to be more lineal and easy to follow.